Fix Customer and SUBSCRIBER_CUSTOMER_KEY relation when a new customer is created directly from Stripe Checkout

arnav13081994 opened the pull request:
<!-- Thank you for helping us out: your contribution means a great deal to the project and the community as a whole! -->


## Description
Due to a race condition in the webhooks `checkout.session.completed` and `customer.created`,  no relation with the related model using the `SUBSCRIBER_CUSTOMER_KEY` was being created. This was because by the time the `update_customer_helper` was invoked in `event_handlers`, the customer didn't exist locally.


<!-- What are you proposing? -->

This PR contains the following changes:

1. Moved `update_customer_helper` from `event_handlers.checkout_webhook_handler` to `Session._attach_objects_post_save_hook`. That way the customer already exists and is associated with the `Session` object along with any metadata.
2. For `Customers` created directly from `Stripe Checkout` that have `SUBSCRIBER_CUSTOMER_KEY` in their metadict, also updated the `Customer` object upstream in Stripe. This is because when `djstripe_sync_models` is run **only** on the `Customer` model, the saved metadict would get wiped off as it doesn't exist or worse may be different in upstream.
3. Updated Corresponding Tests

Checklist:

- [X] I've updated the `tests` or confirm that my change doesn't require any updates.
- [X] I've updated the `documentation` or confirm that my change doesn't require any updates.
- [X] I confirm that my change doesn't drop code coverage below the current level.
- [X] I've updated `migrations` or confirm that my change doesn't make changes to any model.

## Rationale

<!-- 
Why does this project need the change you're proposing? 
If this pull request fixes an open issue, don't forget to link it with `Fix #NNNN` 
-->

All `Customers` objects either created directly using `dj-stripe` or via `Stripe Checkout` will have the same metadata dict and be related to the model referenced using `SUBSCRIBER_CUSTOMER_KEY`

arnav13081994 committed:
Moved update_customer_helper from checkout webhook to its _attach_objects_post_save_hook

This will allow to update the metadata dict in the upstream Customer object that was causing metadata dict sync issues for Customers created directly using Stripe Checkout.

arnav13081994 committed:
Updated Corresponding Tests

jleclanche commented on djstripe/models/checkout.py:
```diff
@@ -101,3 +103,34 @@ class Session(StripeModel):
+                metadata,
+                self.customer.id,
+                djstripe_settings.SUBSCRIBER_CUSTOMER_KEY,
+            )
```
Let's rewrite this like this:

```py
if self.customer and self.metadata:
    key = djstripe_settings.SUBSCRIBER_CUSTOMER_KEY
    current_value = self.metadata.get(key)
    if current_value:
        update_customer_helper({key: current_value}, self.customer.id, key)
```

```

arnav13081994 replied:
Sure. That's definitely more readable.

arnav13081994 committed:
Added 2 pytest fixtures to make tests more manageable

1 fixture will make it easy to create a standard user. And another fixture will make it easy to create a Fake Customer.
